### PR TITLE
Implement PagedAttention V2

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR implements the first part of the PagedAttention V2 kernel, which uses sequence-level parallelism for better work partitioning. Compared to V1, the V2 kernel achieves huge speedup when the batch size is small (e.g., <= 8). We will further optimize the kernel henceforth.

### CODE REVIEW

File: tests/kernels/test_attention.py [hunk 1]
```diff
@@ -96,16 +97,17 @@ def ref_single_query_cached_kv_attention(
         output[i].copy_(out, non_blocking=True)
 
 
+@pytest.mark.parametrize("use_v2", [True, False])
```
Comment (zhuohan123) [inline tests/kernels/test_attention.py:100]:
```suggestion
@pytest.mark.parametrize("version", ["v1", "v2"])
```
And the code below can be changed accordingly

File: vllm/model_executor/layers/attention.py [hunk 2]
```diff
@@ -150,19 +152,62 @@ def single_query_cached_kv_attention(
...
+        max_num_partitions = (
+            (input_metadata.max_context_len + _PARTITION_SIZE - 1) //
+            _PARTITION_SIZE)
+        # NOTE(woosuk): We use a simple heuristic to decide whether to use
+        # PagedAttention V1 or V2. If the number of partitions is 1, we use
+        # V1 to avoid the overhead of reduction. Also, if the number of
+        # sequences or heads is large, we use V1 since there is enough work
+        # to parallelize.
+        # TODO(woosuk): Tune this heuristic.
+        use_v1 = max_num_partitions == 1 or num_seqs * num_heads > 512
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention.py:165]: Why is the threshold 512? Is this number related to the number of SMs a GPU has?

**Review by zhuohan123 — COMMENTED** [review]
Thanks for the great work! In general LGTM. Left some style comments.

File: benchmarks/kernels/benchmark_paged_attention.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+from vllm import attention_ops
+
+NUM_BLOCKS = 1024
+PARTITION_SIZE = 512
+
+
+@torch.inference_mode()
+def main(
+    version: int,
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply benchmarks/kernels/benchmark_paged_attention.py]: Great idea! Fixed.

File: benchmarks/kernels/benchmark_paged_attention.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            size=(num_seqs, num_query_heads, num_partitions),
+            dtype=torch.float32,
+            device=output.device,
+        )
+        max_logits = torch.empty_like(exp_sums)
+
+    def run_benchmark(num_iters: int, profile: bool = False) -> float:
+        if profile:
+            torch.cuda.cudart().cudaProfilerStart()
+        start_time = time.perf_counter()
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply benchmarks/kernels/benchmark_paged_attention.py]: Good catch. Fixed.

File: benchmarks/kernels/benchmark_paged_attention.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+        end_time = time.perf_counter()
+        if profile:
+            torch.cuda.cudart().cudaProfilerStart()
+        return (end_time - start_time) / num_iters
+
+    # Warmup.
+    print("Warming up...")
+    run_benchmark(num_iters=3, profile=False)
+    torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply benchmarks/kernels/benchmark_paged_attention.py]: Fixed!

File: csrc/attention/attention_kernels.cu [hunk 6]
```diff
@@ -85,10 +90,33 @@ __global__ void single_query_cached_kv_attention_kernel(
...
+    return;
+  }
+
+  const int num_context_blocks = DIVIDE_ROUND_UP(context_len, BLOCK_SIZE);
+  const int num_blocks_per_partition = USE_PARTITIONING ? PARTITION_SIZE / BLOCK_SIZE : num_context_blocks;
+
+  // [start_block_idx, end_block_idx) is the range of blocks to process.
+  const int start_block_idx = USE_PARTITIONING ? partition_idx * num_blocks_per_partition : 0;
+  int end_block_idx = MIN(start_block_idx + num_blocks_per_partition, num_context_blocks);
+  int num_blocks = end_block_idx - start_block_idx;
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu]: Good catch! Fixed.

File: csrc/attention/attention_kernels.cu [hunk 7]
```diff
@@ -85,10 +90,33 @@ __global__ void single_query_cached_kv_attention_kernel(
...
+
+  // [start_token_idx, end_token_idx) is the range of tokens to process.
+  int start_token_idx = start_block_idx * BLOCK_SIZE;
+  int end_token_idx = MIN(start_token_idx + num_blocks * BLOCK_SIZE, context_len);
+  int num_tokens = end_token_idx - start_token_idx;
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu]: Fixed. Thanks!

File: csrc/attention/attention_kernels.cu [hunk 8]
```diff
@@ -338,13 +376,167 @@ __global__ void single_query_cached_kv_attention_kernel(
...
+      acc += to_float(tmp_out_ptr[j * HEAD_SIZE + i]) * shared_exp_sums[j] * inv_global_exp_sum;
+    }
+    from_float(out_ptr[i], acc);
+  }
+}
+
 } // namespace vllm
 
-#define LAUNCH_ATTENTION_KERNEL(T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS)                        \
+#define LAUNCH_PAGED_ATTENTION_V1(T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS)                      \
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu]: Great catch! Fixed.

File: csrc/attention/attention_kernels.cu [hunk 9]
```diff
@@ -401,55 +593,219 @@ void single_query_cached_kv_attention_launcher(
...
+  } else if (query.dtype() == at::ScalarType::Half) {
+    CALL_V1_LAUNCHER_BLOCK_SIZE(uint16_t);
+  } else if (query.dtype() == at::ScalarType::BFloat16) {
+    CALL_V1_LAUNCHER_BLOCK_SIZE(__nv_bfloat16);
+  } else {
+    TORCH_CHECK(false, "Unsupported data type: ", query.dtype());
+  }
+}
+
+#define LAUNCH_PAGED_ATTENTION_V2(T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS, PARTITION_SIZE)      \
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu]: Fixed!

File: vllm/model_executor/layers/attention.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention.py:165]: Yes. As we discussed offline, this is a simple heuristic to make sure that the V1 kernel is used when `num_seq * num_heads` is roughly larger than `4 * SM count` in A100 and H100 GPUs. Actually, this can be improved by considering the GPU's actual SM counts. For now, I leave this as future work.

File: csrc/attention/attention_kernels.cu [hunk 10]
```diff
@@ -401,55 +593,219 @@ void single_query_cached_kv_attention_launcher(
...
+  int max_num_partitions = DIVIDE_ROUND_UP(max_context_len, PARTITION_SIZE);
+  int logits_size = PARTITION_SIZE * sizeof(float);
+  int outputs_size = (NUM_WARPS / 2) * head_size * sizeof(float);
+
+  // For paged attention v2 kernel.
+  dim3 grid(num_heads, num_seqs, max_num_partitions);
+  int shared_mem_size = std::max(logits_size, outputs_size);
+  // For paged attention v2 reduce kernel.
+  dim3 grid2(num_heads, num_seqs);
+  int shared_mem_size2 = 2 * max_num_partitions * sizeof(float);
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu]: Good point. Fixed!

File: tests/kernels/test_attention.py (hunk 1 above)
Comment (WoosukKwon) [reply tests/kernels/test_attention.py]: Fixed!

File: csrc/attention/attention_kernels.cu [hunk 11]
```diff
@@ -401,55 +593,219 @@ void single_query_cached_kv_attention_launcher(
...
+  } else if (query.dtype() == at::ScalarType::BFloat16) {
+    CALL_V1_LAUNCHER_BLOCK_SIZE(__nv_bfloat16);
+  } else {
+    TORCH_CHECK(false, "Unsupported data type: ", query.dtype());
+  }
+}
+
+#define LAUNCH_PAGED_ATTENTION_V2(T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS, PARTITION_SIZE)      \
+  vllm::paged_attention_v2_kernel<T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS, PARTITION_SIZE>      \
+  <<<grid, block, shared_mem_size, stream>>>(                                                 \
```
Comment (Yard1) [inline csrc/attention/attention_kernels.cu:690]: do we not need to set `cudaFuncAttributeMaxDynamicSharedMemorySize` here like we do for v1?

Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu:690]: Yes it's not necessary because in V2 each thread block only handles `PARTITION` (=512) tokens. So, if we actually use V2 in all cases, we can remove the shared memory check and support (almost) arbitrary length in all GPUs.

Comment (Yard1) [reply csrc/attention/attention_kernels.cu:690]: Awesome, thanks for explaning!

**Comment by WoosukKwon** [comment]
@zhuohan123 I addressed your comments. PTAL.

File: csrc/attention/attention_kernels.cu (hunk 11 above)
Comment (WoosukKwon) [reply csrc/attention/attention_kernels.cu:690]: @Yard1 That's a good idea! Let's do it in a followup PR.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the awesome work!
